tools/composefs-info.c

```diff
@@ -0,0 +1,294 @@
+   the Free Software Foundation; either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#define _GNU_SOURCE
+
+#include "config.h"
+
+#include "libcomposefs/lcfs-writer.h"
+#include "libcomposefs/lcfs-utils.h"
+#include "libcomposefs/lcfs-internal.h"
+#include "libcomposefs/hash.h"
+
+#include <stdio.h>
+#include <string.h>
+#include <err.h>
+#include <error.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <inttypes.h>
+#include <ctype.h>
+
+#define ESCAPE_STANDARD 0
+#define NOESCAPE_SPACE (1 << 0)
+#define ESCAPE_EQUAL (1 << 1)
+#define ESCAPE_LONE_DASH (1 << 2)
+
+static void print_escaped(const char *val, ssize_t len, int escape)
+{
+	bool noescape_space = (escape & NOESCAPE_SPACE) != 0;
+	bool escape_equal = (escape & ESCAPE_EQUAL) != 0;
+	bool escape_lone_dash = (escape & ESCAPE_LONE_DASH) != 0;
+
+	if (len < 0)
+		len = strlen(val);
+
+	if (escape_lone_dash && len == 1 && val[0] == '-') {
+		printf("\\x%.2x", val[0]);
+		return;
+	}
+
+	for (size_t i = 0; i < len; i++) {
+		uint8_t c = val[i];
+		bool hex_escape = false;
+		const char *special = NULL;
+		switch (c) {
+		case '\\':
+			special = "\\\\";
+			break;
+		case '\n':
+			special = "\\n";
+			break;
+		case '\r':
+			special = "\\r";
+			break;
+		case '\t':
+			special = "\\t";
+			break;
+		case '=':
+			hex_escape = escape_equal;
+			break;
+		default:
+			if (noescape_space)
+				hex_escape = !isprint(c);
+			else
+				hex_escape = !isgraph(c);
+			break;
+		}
+
+		if (special != NULL)
+			printf("%s", special);
+		else if (hex_escape)
+			printf("\\x%.2x", c);
+		else
+			printf("%c", c);
+	}
+}
+
+static void print_node(struct lcfs_node_s *node, char *parent_path)
+{
+	for (size_t i = 0; i < lcfs_node_get_n_children(node); i++) {
+		struct lcfs_node_s *child = lcfs_node_get_child(node, i);
+		cleanup_free char *path = NULL;
+
+		asprintf(&path, "%s/%s", parent_path, lcfs_node_get_name(child));
+
+		uint32_t mode = lcfs_node_get_mode(child);
+		uint32_t type = mode & S_IFMT;
+		const char *payload = lcfs_node_get_payload(child);
+
+		print_escaped(path, -1, NOESCAPE_SPACE);
+
+		if (type == S_IFDIR) {
+			printf("/\t");
+		} else if (type == S_IFLNK) {
+			printf("\t-> ");
+			print_escaped(payload, -1, ESCAPE_STANDARD);
+		} else if (type == S_IFREG && payload) {
+			printf("\t@ ");
+			print_escaped(payload, -1, ESCAPE_STANDARD);
+		}
+		printf("\n");
+
+		print_node(child, path);
+	}
+}
+
+static void digest_to_string(const uint8_t *csum, char *buf)
+{
+	static const char hexchars[] = "0123456789abcdef";
```

Comment:
As far as I can tell, this would literally be the *fourth* copy of this in tree...

Reply:
ooops :)